src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -3433,11 +3485,32 @@ StartPlacementExecutionOnSession(TaskPlacementExecution *placementExecution,
 		ExtractParametersForRemoteExecution(paramListInfo, &parameterTypes,
 											&parameterValues);
 		querySent = SendRemoteCommandParams(connection, queryString, parameterCount,
-											parameterTypes, parameterValues);
+											parameterTypes, parameterValues,
+											binaryResults);
 	}
 	else
 	{
-		querySent = SendRemoteCommand(connection, queryString);
+		/*
+		 * We only need to use SendRemoteCommandParams when we desire
+		 * binaryResults. One downside of SendRemoteCommandParams is that it
+		 * only supports one query in the query string. In some cases we have
+		 * more than one query. In those cases we already make sure before that
+		 * binaryResults is false.
+		 *
+		 * XXX: It also seems that SendRemoteCommandParams does something
+		 * strange/incorrectly with select statements. In
+		 * isolation_select_vs_all.spec, when doing an s1-router-select in one
+		 * session blocked an s2-ddl-create-index-concurrently in another.
```

Comment:
could the issue be the CREATE INDEX CONCURRENTLY statement?

Reply:
I just tried using only binary protocol for the SELECT and it still got blocked. This actually seems like a problem in our deadlock detection too, since it does not report the blocking using dump_global_wait_edges. The lock that it's waiting on is actually the `virtualxid` lock. From what I can tell from reading the documentation on CREATE INDEX CONCURRENTLY is that it waits for any read only statements that have not finished and started before starting CREATE INDEX CONCURRENTLY. So it seems like for some reason `SendRemoteCommandParams` does not mark the statement as finished (or Postgres handles it wrongly).

Reply:
Interesting. Based on documentation waiting kind of makes sense since the transaction could still make modifications, but since it's in read committed mode it would use a new snapshot so not waiting is also ok. Not sure why the protocol matters, seems like a pg issue.

Reply:
cross-reference https://github.com/citusdata/citus/pull/5910#discussion_r858416884